// KT-442 Type inference fails on with()

fun <T> funny(f : fun() : T) : T = f()

fun testFunny() {
    val <!UNUSED_VARIABLE!>a<!> : Int = funny {1}
}

fun <T> funny2(f : fun(t : T) : T) : T <!NO_RETURN_IN_FUNCTION_WITH_BLOCK_BODY!>{}<!>

fun testFunny2() {
    val <!UNUSED_VARIABLE!>a<!> : Int = funny2 {it}
}

fun box() : String {
    return generic_invoker { it }
}

fun <T> generic_invoker(gen : fun (String) : T) : T {
    return gen("")
}

fun <T> T.with(f : fun T.()) {
    f()
}

fun main(args : Array<String>) {
    val <!UNUSED_VARIABLE!>a<!> = 1 with {
        plus(1)
    }
}